### ISSUE TITLE
Swift Target

### ISSUE THREAD

**Issue by jeffreyguenther** [antlr/antlr4#945]
I did a quick search and I didn't see anything written about this yet. What's the likelihood of a Swift target for ANTLR?

There are C#, Javascript, and Python targets at the moment.

What does it take to implement a target? Given that Swift is more Java-like, it seems like it should be possible. Maybe start with a code translator if there is one for (Java to Swift), and iterate towards a more idiomatic implementation.

**Comment by parrt** [comment]
Hi. We have no current plans although I'm hoping to learn Swift much better than I do now so I might end up doing it eventually. Building a target is fairly involved but the guy who did many of the targets managed to pull off Python in about a month just by tweaking the code until the unit tests get the same output :)

**Comment by jrguenther** [comment]
Good to know.

What are the hardest parts of implementing a target?

**Comment by parrt** [comment]
Maybe the tedium. ;)  I think it's the initial learning curve of understanding what the hell needs to be done; and nothing is written up.  If you're serious about it, I can get you crackin via Skype sometime.  I'd suggest manual not auto translation to start.

**Comment by parrt** [comment]
I should also say that the hundreds of auto-created runtime tests for each target is very helpful.

**Comment by EpsilonPrime** [comment]
Right now I'd complain about the difficulties of getting it to output both header files and language files, but I should have that tackled before you get there.

**Comment by parrt** [comment]
hahaha. Want me to look at that?

**Comment by EpsilonPrime** [comment]
I'm close (it looks like some of the code has moved along like adding support for recognizers and vistors since it was put in there).  If I'm still having troubles tomorrow I'll ping you.

**Comment by EpsilonPrime** [comment]
And some amount of success.  Turns out setting some string templates to something besides null helps.  :)

Some changes to tool/src/org/antlr/v4/codegen/CodeGenerator.java were necessary to get this to work but it's not ready yet for a pull request (and I'm only 1000 commits behind!):
https://github.com/antlr/antlr4/compare/master...davesisson:feature/templates

**Comment by parrt** [comment]
cool.

**Comment by jrguenther** [comment]
I'm definitely interested in doing it. It's this or implement OpenGL integration for JavaFX. I want to integrate my language into graphics application and OpenGL in JavaFX right now is not really possible, so this seems much easier to do. And I think the Swift community will benefit from it a lot. As far as I've found [PEGKit](https://github.com/itod/pegkit) and the developer's older library ParseKit are the only options for easy to use parser tools. There might be others, but I haven't found them yet.

I have another commitment at the moment, so I'll need to chip away at it over time.

@EpsilonPrime are you working on a Swift target right now?

**Comment by parrt** [comment]
He's working on C++

**Comment by jrguenther** [comment]
Sounds great. Thanks @parrt! I appreciate all your hard work in giving us a great framework. :smile:

**Comment by gorums** [comment]
Anybody is working in this ticket?

**Comment by parrt** [comment]
Not that I know of.

**Comment by jrguenther** [comment]
I'm still tied up with other things at the moment, so if you keen to start, go for it. I'll join in the effort when I'm free.

**Comment by janyou** [comment]
https://github.com/janyou/Antlr-Swift-Runtime
https://github.com/janyou/ANTLR-Swift-Target

**Comment by parrt** [comment]
heh, cool! how complete and tested is the runtime?

**Comment by janyou** [comment]
I use it for translating other languages to Swift. I think more testing is needed.

**Comment by parrt** [comment]
Ok, have you seen [Creating an ANTLR Language Target](https://github.com/antlr/antlr4/blob/master/doc/creating-a-language-target.md)?  It explains the testing mechanism.

**Comment by ericvergnaud** [comment]
Hi,

great that you have created this target, but can you elaborate in the use case?
you dont need antlr to produce code, only to parse code
you mention below that you use the target for translating from other languages to Swift, what role does the Swift target play in that scenario?

Eric

**Comment by janyou** [comment]
> Ok, have you seen Creating an ANTLR Language Target? It explains the testing mechanism.

OK,I will make the test rig.

> great that you have created this target, but can you elaborate in the use case?
> you dont need antlr to produce code, only to parse code
> you mention below that you use the target for translating from other languages to Swift, what role does the Swift 
> target play in that scenario?

producing code is the necessary condition of parsing code.
I need a translation tools with convenience UI.
Java isn't good at UI. I try antlr Javascript target,but it is too slow.
So I translated Antlr Java target to Swift target,it works well and as fast as Java version.

**Comment by janyou** [comment]
All test cases in runtime-testsuite have passed.(modifying a few template stg files)

**Comment by ericvergnaud** [comment]
Hi,

Great! 
I’m not seeing those tests in Travis-CI, though…
Can you provide links?

Eric

**Comment by janyou** [comment]
Sorry, never create it .  I'll do it later

**Comment by parrt** [comment]
fantastic. maybe I should just make a branch and try to install and test?

**Comment by janyou** [comment]
It's difficult to merge MultipleAlternativesWithCommonLabel.stg and TokenAndRuleContextString.stg.
Let me try to get a handle on this.

**Comment by ericvergnaud** [comment]
maybe add a target template for "optional value":

<Cast("BinaryContext","$ctx")><OptionalValue(."e(0)")>.v * <Cast("BinaryContext","$ctx")>.<OptionalValue(."e(1)")>.v

where OptionalValue(value) ::= <<<value>!>> fo swift and simply <<<value>>> for other targets

**Comment by janyou** [comment]
good idea!
but must modify other target's  *.test.stg .

**Comment by parrt** [comment]
Any movement on the swift target guys?

**Comment by slcott** [comment]
I'm really interested in the Swift target. 
Is there an expectation when it will be available?

**Comment by parrt** [comment]
i need to bump it up in priority before I can work it in ;)

**Comment by deus777** [comment]
hanging out for this one too

**Comment by parrt** [comment]
Working on a paper then have to prep classes. hopefully time will pop up soon.

**Comment by slcott** [comment]
Running into some issues using the swift target.
https://github.com/janyou/ANTLR-Swift-Target/issues/3#issuecomment-216066145

**Comment by bzeiss** [comment]
For what it's worth: while I probably won't start to implement the target, I'm pretty sure that I would use it without a doubt when its done :-)

**Comment by parrt** [comment]
heads up. I'm hoping to look at this next week. @hanjoes any progress?

**Comment by hanjoes** [comment]
Took several days off in the past two weeks, I've setup the project and tweaked the code though, planning to finish it before next Wed. (could be faster)

**Comment by parrt** [comment]
ok, cool. I'm playing with plugin at moment. better icons, darcula colors.

**Comment by hanjoes** [comment]
nice! :)

**Comment by hanjoes** [comment]
when I tried to integrate the change with swift target, it seems the newer version of the swift target has added force unwrap ("!") to all attribute decls, in Swift.stg:

`AttributeDecl(d) ::= "<d.name>: <SwiftTypeMap.(d.type)><if(d.initValue)> = <d.initValue><else>!<endif>"
`
**note the exclamation mark after "\<else\>" tag.**

So if I wrote grammar:

```
grammar Test;

classDef returns [scope: XClass?]
	: 'class'
	;
```

I expect to have a field like:
`public var scope: XClass?`

But now because of the "forced" force unwrap, what i get is: 
`public scope: XClass?!`

It compiles but seems a little redundant and I think it would be better to let user decide whether the attribute is an optional or not.

**Comment by parrt** [comment]
Interesting.  Yeah, seems like `!` should be at the discretion of the programmer, unless the actions that refer to them can't figure out what to do. Maybe that's it. referencing $classDef.scope might require forced unwrap?

**Comment by hanjoes** [comment]
I presume actions are all user defined operations? If it's the case users should always know whether $classDef.scope ought be an optional or not. The compiler/runtime are doing quite strict checks for optionals.

**Comment by parrt** [comment]
True! `{$classDef.scope}` would become plain `classDef.scope` and `{$classDef.scope!}` would be `classDef.scope!`  Thoughts @jrguenther?

**Comment by hanjoes** [comment]
@jrguenther please refer to [this link](https://github.com/antlr/antlr4/blob/master/doc/parser-rules.md#rule-attribute-definitions), we were discussing whether the "rule attributes" are optionals or not should be totally controlled by the programmer.

**Comment by jrguenther** [comment]
Ah, yes, I forgot about that use of rule attributes. I don't use actions when I use ANTLR.

I agree with allowing the user to choose if the attribute is to be treated as an optional and then when to unwrap the optional. This does raise a question in my mind Java that now has `Optional`s as well as of Java 8. Should this be available in the Java target too? How does it fit in with Terrence's original design? How much variation is tolerated among targets?

**Comment by parrt** [comment]
I probably won’t be doing Optional anytime soon as it’s not a  high priority. variation is tolerated to make the target behave more like programmers in that language would expect :)

Ter

**Comment by lspicer** [comment]
I was just curious what the current state is for this development. Is there anything documenting the progress? I'm potentially interested in contributing or helping out if I can, but it's not clear where the work is being done, or what still needs to be done. Thanks and keep up the good work!

**Comment by jeffreyguenther** [comment]
+1 to @lspicer's question. @hanjoes and @janyou how is your work coming along? Is there anything we, the community, can do to help?

I have time I can devote to working on the target. I'm not a Swift or XCode expert, but I'm keen to see this target completed as I have a project that'd I'd like to do with it. 😄 So I'm willing to chip in to do whatever is necessary.

Once the Swift target is built, there will be a few things we need to do to make ANTLR as friendly to the Swift community is the Java community:

- Prepare documentation on how to use the ANTLR tool alongside XCode. I'd love to see this workflow as easy as adding a gradle task to generate source from a grammar file.

- Prepare the code to work with the XCode package managers (Carthage, CocoaPods, and Alcatraz are the one's I'm aware of.) I believe the Swift package manager was released with Swift 3, but I'm not aware of it using a central repository yet. Maybe someone else knows better?

What do you say we push this Swift target forward so the next time Terence releases ANTLR, the Swift target is in it?

**Comment by hanjoes** [comment]
I would also happy to see the Swift target got shipped with the future ANTLR releases. 

@janyou was the one getting most of the work done for the current Swift target. 
I worked on a small patch to enable ANTLR to support language idioms like those in Swift.

 I think the we might need @parrt to decide how/when the Swift target is merged into the ANTLR mainline.

**Comment by jeffreyguenther** [comment]
Ya, Terence will be the one who decides. I think there is some work we can do in the meantime to make it easy for him to accept the pull request.

Is your patch a nice-to-have or something required for the Swift target? The reason I ask is because building the target is a big chunk of work and getting the patch in place might be something done best as enhancement after the first version of the target is in place, especially if it means a change to ANTLR itself. @parrt will be able to say what's in scope.

@janyou's code is not organized in quite the same way the ANTLR codebase requires. Reorganizing the code so it follows the convention is, in my view, the first thing that needs to happen. Then we need to make sure that the tests are properly generated. Instructions can be found in the [language target docs](https://github.com/antlr/antlr4/blob/master/doc/creating-a-language-target.md). I think we need to do this work first.

**Comment by jeffreyguenther** [comment]
Also, I noticed that janyou's implementation changes the ATN serialization to JSON. I'm not sure that's going to fly with Terence as it's not a change to make the target more idiomatic.

**Comment by hanjoes** [comment]
I agree with you. 

I tried to use @janyou 's implementation but got stuck on the generated code, since right now ANTLR doesn't seem to support language idioms that put type decl after the variable definition.

But anyways, it takes time to re-organize/re-write the current Swift implementation and my little patch will no way be the blocker. So it's a good idea to push the process forward.

**Comment by jeffreyguenther** [comment]
Hmm. That doesn't seem quite right that the wrong code is output. Unless I'm missing something, we should be able to write the Swift code in whatever way we want using StringTemplate. As long as the right data is available, we can reformat it however we need to generate the Swift code.

In any case, we'll tackle it when it comes up. Next, we need to decide where to do this work. I think it makes the most sense to work on a common repo that we issue a single PR from when we're ready. I've added you as a collaborator on my fork and we can work there.

I'm just wrapping up creating the correct folders and stub files.

**Comment by jeffreyguenther** [comment]
@janyou do you have any objection with @hanjoes and I using your code as the foundation for our effort? I'd like to avoid us duplicating the work you've already completed.

**Comment by jeffreyguenther** [comment]
Did a little research this evening:

Swift [uses an extended version of CommonMark for documentation](http://useyourloaf.com/blog/swift-documentation-quick-guide/)
If we use a tool called [jazzy](https://github.com/realm/jazzy), we'll generate docs in a similar style to Apple's developer docs.

**Comment by jeffreyguenther** [comment]
@parrt I think I'm ready for that Skype overview if you're still willing to give it.

**Comment by janyou** [comment]
I welcome you to contribute to the ANTLR Swift Target project and You are free to use my code.
I also want someone to finish the merge.

The main job is to merge the test templates. The following points for reference:

1.CompositeParsers/DelegatorRuleOverridesLookaheadInDelegate.stg

**Other  target**
```
decl : type_ ID ';'
  | type_ ID init ';' {<write("\"JavaDecl: \" + $text")>};
    init : '=' INT;
```

**Swift target**
```
decl : type_ ID ';'
		| type_ ID initIt ';' {<write("\"JavaDecl: \" + $text")>};
	initIt : '=' INT;
```

**Difference**:
`init` -> `initIt`
note:  `init` is a reserved word in Swift.

2.LeftRecursion/MultipleAlternativesWithCommonLabel.stg

**Other  target**
```
	e returns [int v]
	  : e '*' e     {$v = <Cast("BinaryContext","$ctx")>.e(0).v * <Cast("BinaryContext","$ctx")>.e(1).v;}  # binary
	  | e '+' e     {$v = <Cast("BinaryContext","$ctx")>.e(0).v + <Cast("BinaryContext","$ctx")>.e(1).v;}  # binary
```
 
**Swift target**
```
	e returns [int v]
	  : e '*' e     {$v = <Cast("BinaryContext","$ctx")>.<OptionalValue("e(0)")>.v * <Cast("BinaryContext","$ctx")>.<OptionalValue("e(1)")>.v;}  # binary
	  | e '+' e     {$v = <Cast("BinaryContext","$ctx")>.<OptionalValue("e(0)")>.v + <Cast("BinaryContext","$ctx")>.<OptionalValue("e(1)")>.v;}  # binary
```	
**Difference**:
`e(0)` ->  `<OptionalValue("e(0)")`
`e(1)` ->  `<OptionalValue("e(1)")`
note: `e(0)` and `e(1)` are optional values.
other targets'.test.stg should be added one line : 

> OptionalValue(value) ::= "\<value\>" 


3.ParserExec/ReferenceToATN.stg

**Other  target**
```
	a : (ID|ATN)* ATN? {<writeln("$text")>} ;
	ID : 'a'..'z'+ ;
	ATN : '0'..'9'+;
	WS : (' '|'\n') -> skip ;
```

**Swift target**
```
	a : (ID|Atn)* Atn? {<writeln("$text")>} ;
	ID : 'a'..'z'+ ;
	Atn : '0'..'9'+;
	WS : (' '|'\n') -> skip ;
```

**Difference**:
`ATN` -> `Atn`
note: `ATN` is Antlr runtime's class name.

**Comment by parrt** [comment]
I'm hoping to get back to antlr stuff Oct/Nov 2016

**Comment by ewanmellor** [comment]
I have created a branch here: https://github.com/ewanmellor/antlr4/tree/swift-target

This has the current antlr/antlr4 master branch, but I've added the changes from @janyou mentioned on Dec 31 2015, from https://github.com/janyou/Antlr-Swift-Runtime and https://github.com/janyou/ANTLR-Swift-Target.

Most tests pass, a few fail.  I'm also sure a few of these changes won't fly as-is, since it looks like they break other backends.  It looks like a good start though.

If anyone wants to work on this, I'd be happy for you to pull my branch, or send me pull requests.

**Comment by jeffreyguenther** [comment]
@ewanmellor Did you keep the JSON ATN serialization in your branch?

**Comment by ewanmellor** [comment]
@jeffreyguenther I just copied janyou's code.  I don't know what ATN serialization is (I'm new to Antlr).  Can you tell me what I should be looking for?

**Comment by jeffreyguenther** [comment]
I think the easiest way to find it is to search for “json” in your editor. janyou tweaked the ATN serialization to output JSON rather than keeping with ANTLR’s normal approach. JSON might be nice to have, but I don’t think Terence will go for the difference from the rest of ANTLR.

ATN stands for augmented transition network <https://theantlrguy.atlassian.net/wiki/display/~admin/2011/04/17/These+are+some+of+my+favorite+things--PDA,+RTN,+ATN,+DFA,+NFA> - think a serialized form of the state machine used to represent the grammar.

**Comment by janyou** [comment]
@ewanmellor Please refer to [Manually build](https://github.com/janyou/ANTLR-Swift-Target/blob/master/ManuallyBuild.md) and [Run Runtime Testsuite](https://github.com/janyou/ANTLR-Swift-Target/blob/master/RuntimeTestsuite.md

**Comment by ewanmellor** [comment]
@janyou Yes, I have copied all those instructions, and applied them to my branch.

**Comment by ewanmellor** [comment]
@jeffreyguenther Yes, I see it.  Yes, that's all still in there from janyou's code.  It'll need entirely rewriting by the looks of it.

I'm not going to get to this for at least a week, probably more.  Patches welcome!

**Comment by jeffreyguenther** [comment]
Thought we could have this discussion over here with all the different language targets being discussed in the new thread. 

What do we need to do to get things ready for @parrt to merge the Swift target?

I think we should work off of @ewanmellor's fork since he's done all the copy and paste to put things in the right place. Let's make a list, divide the work, and finish this up.

**Comment by janyou** [comment]
@sharwell's recommendations：

1.modify  MultipleAlternativesWithCommonLabel.stg
rename `OptionalValue`  to ?

2.modify Swift.test.stg
```
IgnoredTests ::= [
	"ParserExec.ReferenceToATN": true,
	"ParserExec.ReferenceToATN_1": true,
	"ParserExec.ReferenceToATN_2": true,
	default: false
]
```

**Comment by ewanmellor** [comment]
**Swift Target**

I have updated the branch at https://github.com/ewanmellor/antlr4/tree/swift-target.  I shall rebase this branch to 4.6-SNAPSHOT.  I have also enabled Travis builds for this branch at https://travis-ci.org/ewanmellor/antlr4.

I have created #1327, #1328, #1329, #1330, #1331 for the issues raised above.

I think I have fixed the issue with renaming ATN (https://github.com/ewanmellor/antlr4/commit/8b0373a580deb801e88f54cc40638cd7e023b1cd) so there won't be a need to disable those tests.

**Comment by jeffreyguenther** [comment]
Nice work @ewanmellor! I was just in the process of doing some of this... Let me know when you finish up the rebase and I'll re-fork to tackle the issues I'm able to solve.

**Comment by ewanmellor** [comment]
@parrt This can be closed now, following the work in #1319.
